[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98826
* #98730
* #99132
* __->__ #99131

**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with @Xia-Weiwen 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

leslie-fang-intel committed:
Enable decomposed.quant/dequant lowering and vec code gen

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1954,6 +1961,14 @@ def to_dtype(x, dtype):
+                                ).data.layout.size.__len__()
+                                == 0
+                            ):
+                                opt_ctx.is_load_int_scalar_tensor = True
```
Who is using this variable?

leslie-fang-intel replied:
It's not used. I will remove it.

leslie-fang-intel replied:
Removed it.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -1652,7 +1652,14 @@ def load(self, name: str, index: sympy.Expr):
+            buffer = V.graph.get_buffer(name)
+            if (
+                load_dtype in [torch.int32, torch.int64]
+                and (buffer.data.layout.size.__len__() == 0)
```
`not buffer.data.layout.size` or `len(buffer.data.layout.size) == 0`

leslie-fang-intel replied:
Thanks for the comments and changed.

leslie-fang-intel committed:
Update on "Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@jansel Could you help to take a look of this PR?

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -3895,6 +3896,44 @@ def _realize(x):
 
 
+# The difference between quantize_per_tensor.default and quantize_per_tensor.tensor is
+# scale and zero_point is scalar or scalar tensor
+@register_lowering(torch.ops.quantized_decomposed.quantize_per_tensor.default)
+def quantize_per_tensor_default(input, scale, zero_point, quant_min, quant_max, dtype):
+    def clamp(v):
+        return maximum(quant_min, minimum(quant_max, v))
+
+    output = clamp(add(round(div(input, scale)), zero_point))
+    output = to_dtype(output, dtype)
+    return output
+
+
+# The difference between dequantize_per_tensor.default and dequantize_per_tensor.tensor is
+# scale and zero_point is scalar or scalar tensor
+@register_lowering(torch.ops.quantized_decomposed.dequantize_per_tensor.default)
+def dequantize_per_tensor_default(
+    input, scale, zero_point, quant_min, quant_max, dtype
+):
+    output = mul(sub(to_dtype(input, torch.float32), zero_point), scale)
+    return output
+
+
+@register_lowering(torch.ops.quantized_decomposed.quantize_per_tensor.tensor)
+def quantize_per_tensor_tensor(input, scale, zero_point, quant_min, quant_max, dtype):
+    def clamp(v):
+        return maximum(quant_min, minimum(quant_max, v))
+
+    output = clamp(add(round(div(input, scale)), zero_point))
+    output = to_dtype(output, dtype)
+    return output
+
+
+@register_lowering(torch.ops.quantized_decomposed.dequantize_per_tensor.tensor)
+def dequantize_per_tensor_tensor(input, scale, zero_point, quant_min, quant_max, dtype):
+    output = mul(sub(to_dtype(input, torch.float32), zero_point), scale)
+    return output
+
+
```
These might be cleaner to write as decomps in `decomposition.py`

leslie-fang-intel replied:
Hi @jansel, thanks for your comments. There is already an existing but un-landed PR https://github.com/pytorch/pytorch/pull/97321 from quantization team to register these operations into `decomposition table`. However, synced with @jerryzh168 previously, quantization team has not decided whether to decompose these operations yet. That's why we do these operations lowering here. Do you think it's ok we do this lowering firstly and remove them after quantization team decide to merge above PR https://github.com/pytorch/pytorch/pull/97321? also cc @jgong5 @EikanWang here.

jansel replied:
If we decide not to merge #97321, you can just put the decomps in the inductor codebase here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/decomposition.py

When you call AotAutograd you get to pick your own decomps, so even if inductor decomposes it others can decide not to.

The above code does the exact same thing as using decomps, just in a slightly hard to read way.

leslie-fang-intel replied:
@jansel Thanks for your suggestions. Move these ops into https://github.com/pytorch/pytorch/blob/main/torch/_inductor/decomposition.py. Please help to take a look again.

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented on torch/_inductor/decomposition.py:
```diff
@@ -35,7 +37,17 @@
     ]
 )
-decompositions = {**core_aten_decompositions(), **inductor_decompositions}
+quantized_decomps = get_decompositions(
```
`get_decompositions()` looks up an existing decomposition registered in `torch._decomp`.  Are there registered decomps for these there? If not I think this will just return an empty dict.

leslie-fang-intel replied:
No, it's not registered there. Remove `quantized_decomps`. Please help to take a check.

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][quant]Enable decomposed.quant/dequant lowering and vec code gen"


**Summary**
Since current quantization flow has not decomposed quant/dequant into prim ops, in this PR

- We enable the quant/dequant decomposition as lowering inside inductor.
- For the `decomposed.quant/dequant.tensor` overload, there are loading of scalar tensor of `zero point` and `scale`, we need to enable the vec code gen for these op overloads.
- Minor change as adding `is_load_uint8_as_float` and `is_store_float_as_uint8` default value `False` into `OptimizationContext`.

**TestPlan**
```
cd test/inductor && python -m pytest test_cpu_repro.py -k test_dequant_quant_lowering
```
co-author with Xia-Weiwen 

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
